### PR TITLE
Buffered insertions

### PR DESCRIPTION
**Pull request by JulienR1**
Adds in-memory batching to the sink.
The pqueue has been removed in favor of a single promise that inserts data while more data is stored locally asynchronously.

The behavior can be configured in `.env` or with the CLI using `MAX_BUFFER_SIZE` and `INSERTION_DELAY`.

### CODE REVIEW

File: src/clickhouse/handleSinkRequest.ts [hunk 1]
```diff
@@ -1,122 +1,158 @@
...
+  ) {
+    const waitTime = nextUpdateTime - new Date().getTime();
+    await new Promise((resolve) => setTimeout(resolve, waitTime));
   }
 
-  // Prevent queue from growing too large
-  prometheus.queue_size.set(queue.size);
-  if (queue.size > config.queueLimit) await setTimeout(1000);
+  if (new Date().getTime() > nextUpdateTime) {
+    await promise;
```
Comment (DenisCarriere) [inline src/clickhouse/handleSinkRequest.ts:52]: p-queue

File: src/clickhouse/handleSinkRequest.ts [hunk 2]
```diff
@@ -1,122 +1,158 @@
...
+
+      if (Object.keys(entityChanges).length > 0) {
+        for (const [table, values] of Object.entries(entityChanges)) {
+          if (values.length > 0) {
+            await client.insert({ table, values, format: "JSONStringsEachRow" });
+          }
+        }
+      }
+
+      resolve();
```
Comment (DenisCarriere) [inline src/clickhouse/handleSinkRequest.ts:102]: pqueue

**Review by DenisCarriere — COMMENTED** [review]
Overall looks good, but your Promise handling looks a bit wonky

You are trying to recreate your Promise based Queue batching

You should simply just bring back in PQueue for that section alone

Have a look at:
https://github.com/sindresorhus/p-queue#advanced-example

```ts
await queue.onEmpty()
// or
await queue.onIdle();
```
To know when all the promises have been resolved to continue to the next step

**Comment by JulienR1** [comment]
Edited the workflow with `p-queue`. It stores 2 promises at once: the timeout promise and the insertion promise.

The timeout promise keeps track of the next insertion time. It has to be in the pqueue to block the process until it is the correct insertion time later on.

The insertion promise is the call to ClickHouse. It will delay everything if it is slow.

Here is a summary of the process:

image

**Comment by DenisCarriere** [comment]
👍  update looks good

**Pull request merged**